autooai.py

```diff
@@ -0,0 +1,142 @@
+import re
+import shutil
+import logging
+from datetime import date
+from datetime import timedelta
+
+import vcr
+import furl
+from lxml import etree
+from dateutil.parser import parse
+
+from scrapi import requests
+
+URL_RE = re.compile(r'(https?:\/\/[^\/]*)')
+
+NAMESPACES = {'dc': 'http://purl.org/dc/elements/1.1/',
+              'oai_dc': 'http://www.openarchives.org/OAI/2.0/',
+              'ns0': 'http://www.openarchives.org/OAI/2.0/'}
+
+BASE_SCHEMA = ['title', 'contributor', 'creator', 'subject',
+               'description', 'language', 'publisher']
+
+OAI_TEMPLATE = """'''
+Harvester for the {0} for the SHARE project
+Example API call: {1}
+'''
+from __future__ import unicode_literals
+
+from scrapi.base import OAIHarvester
+
+
+class {2}Harvester(OAIHarvester):
+    short_name = '{3}'
+    long_name = '{4}'
+    url = '{5}'
+
+    base_url = '{6}'
+    property_list = {7}
+    timezone_granularity = {8}
+"""
+
+logger = logging.getLogger(__name__)
+
+
+def get_oai_properties(base_url, shortname, start_date, end_date):
+    """ Makes a request to the provided base URL for the list of properties
+        returns a dict with list of properties
+    """
+
+    try:
+        prop_base = furl.furl(base_url)
+        prop_base.args['verb'] = 'ListRecords'
+        prop_base.args['metadataPrefix'] = 'oai_dc'
+
+        pre_names = push_request_boundaries(shortname, prop_base, start_date, end_date)
+        all_names = [name.tag.replace('{' + NAMESPACES['dc'] + '}', '') for name in pre_names]
+        return list({name for name in all_names if name not in BASE_SCHEMA}) + ['setSpec']
+
+    # If anything at all goes wrong, just render a blank form...
+    except Exception as e:
+        raise ValueError('OAI Processing Error - {}'.format(e))
+
+
+def push_request_boundaries(shortname, prop_base, startdate, enddate):
```

Comment:
While this function is awesome, we don't really need it if we're gonna get rid of date ranges when querying for elements to add to the properties list, maybe.